<a href="https://colab.research.google.com/github/Jojomaque/Portfolio/blob/main/Jojo_Preparing_data_for_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Preparing Data for Analysis**



*   Exercise 1 - Checking for null vlues
*   Exercise 2 - Removing null values
*   Exercise 3 - Removing duplicate records from a data set
*   Exercise 4 - Normalising values where necessary
*   Exercise 5 - Correcting the format of dates where necessary









### Exercise 1 - check for null values 
---

1. Read data from the file housing_in_london_yearly_variables.csv from this link: https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv 
2. check if any NA values exist in the dataframe and print the result 
3. use df.info() to see which columns have null entries (*Hint: if the non-null count is less than total entries, column contains missing/NA entries*)  

**Test output**:
True
.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071) 



In [1]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv')
housing_london_original = df
housing_london_original.isnull().values.any()

True

In [2]:
#.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071)
housing_london_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1071 entries, 0 to 1070
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               1071 non-null   object 
 1   area               1071 non-null   object 
 2   date               1071 non-null   object 
 3   median_salary      1049 non-null   float64
 4   life_satisfaction  352 non-null    float64
 5   mean_salary        1071 non-null   object 
 6   recycling_pct      860 non-null    object 
 7   population_size    1018 non-null   float64
 8   number_of_jobs     931 non-null    float64
 9   area_size          666 non-null    float64
 10  no_of_houses       666 non-null    float64
 11  borough_flag       1071 non-null   int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 100.5+ KB


### Exercise 2 - remove null values 
---

1. Remove rows with NA values for `life_satisfaction` (use [ ] even if only one column in list)
2. Remove all NA values across whole dataframe 

**Test output**:  
1.  Row count reduced to 352 rows
2.  Row count reduced to 267 rows

In [3]:
#Remove rows with NA values for life_satisfaction
housing_london_original.dropna(subset = ['life_satisfaction'])

code                  area  ... no_of_houses  borough_flag
613   E09000002  barking and dagenham  ...      71079.0             1
614   E09000003                barnet  ...     139346.0             1
615   E09000004                bexley  ...      95037.0             1
616   E09000005                 brent  ...     112083.0             1
617   E09000006               bromley  ...     135036.0             1
...         ...                   ...  ...          ...           ...
1010  E12000009            south west  ...          NaN             0
1013  E92000001               england  ...   24172166.0             0
1017  N92000002      northern ireland  ...          NaN             0
1018  S92000003              scotland  ...          NaN             0
1019  W92000004                 wales  ...          NaN             0

[352 rows x 12 columns]

In [4]:
#Remove rows with NA values for whola dataframe
housing_london_original.dropna()

code                  area  ... no_of_houses  borough_flag
613   E09000002  barking and dagenham  ...      71079.0             1
614   E09000003                barnet  ...     139346.0             1
615   E09000004                bexley  ...      95037.0             1
616   E09000005                 brent  ...     112083.0             1
617   E09000006               bromley  ...     135036.0             1
...         ...                   ...  ...          ...           ...
999   E09000031        waltham forest  ...     103029.0             1
1000  E09000032            wandsworth  ...     146162.0             1
1001  E09000033           westminster  ...     124509.0             1
1008  E12000007                london  ...    3556161.0             0
1013  E92000001               england  ...   24172166.0             0

[267 rows x 12 columns]

## Dropping duplicates
---

* To remove duplicate rows based on duplication of values in all columns  
  `df.drop_duplicates()`  

* To remove rows that have duplicate entries in a specified column  
  `df.drop_duplicates(subset = ['Make'])`  

* To remove rows that have duplicate entries in multiple columns  
  `df.drop_duplicates(subset = ['Make', 'Model'])` 

* Remove duplicate rows keeping the last instance rather than the first (default):  
  `df.drop_duplicates(keep='last')`  

### Exercise 3 - Removing duplicate entries 
---

remove duplicate `area` entries keeping first instance  

**Test output**:  
 Dataframe now contains 50 rows all with date 1999-12-*01* 

In [5]:
#remove duplicate area entries keeping first instance
housing_london_original.drop_duplicates(subset = ['area'])


code                      area  ... no_of_houses  borough_flag
0   E09000001            city of london  ...          NaN             1
1   E09000002      barking and dagenham  ...          NaN             1
2   E09000003                    barnet  ...          NaN             1
3   E09000004                    bexley  ...          NaN             1
4   E09000005                     brent  ...          NaN             1
5   E09000006                   bromley  ...          NaN             1
6   E09000007                    camden  ...          NaN             1
7   E09000008                   croydon  ...          NaN             1
8   E09000009                    ealing  ...          NaN             1
9   E09000010                   enfield  ...          NaN             1
10  E09000011                 greenwich  ...          NaN             1
11  E09000012                   hackney  ...          NaN             1
12  E09000013    hammersmith and fulham  ...          NaN             1
13  E09000014                  haringey  ...          NaN             1
14  E09000015                    harrow  ...          NaN             1
15  E09000016                  havering  ...          NaN             1
16  E09000017                hillingdon  ...          NaN             1
17  E09000018                  hounslow  ...          NaN             1
18  E09000019                 islington  ...          NaN             1
19  E09000020    kensington and chelsea  ...          NaN             1
20  E09000021      kingston upon thames  ...          NaN             1
21  E09000022                   lambeth  ...          NaN             1
22  E09000023                  lewisham  ...          NaN             1
23  E09000024                    merton  ...          NaN             1
24  E09000025                    newham  ...          NaN             1
25  E09000026                 redbridge  ...          NaN             1
26  E09000027      richmond upon thames  ...          NaN             1
27  E09000028                 southwark  ...          NaN             1
28  E09000029                    sutton  ...          NaN             1
29  E09000030             tower hamlets  ...          NaN             1
30  E09000031            waltham forest  ...          NaN             1
31  E09000032                wandsworth  ...          NaN             1
32  E09000033               westminster  ...          NaN             1
33  E12000001                north east  ...          NaN             0
34  E12000002                north west  ...          NaN             0
35  E12000003  yorkshire and the humber  ...          NaN             0
36  E12000004             east midlands  ...          NaN             0
37  E12000005             west midlands  ...          NaN             0
38  E12000006                      east  ...          NaN             0
39  E12000007                    london  ...          NaN             0
40  E12000008                south east  ...          NaN             0
41  E12000009                south west  ...          NaN             0
42  E13000001              inner london  ...          NaN             0
43  E13000002              outer london  ...          NaN             0
44  E92000001                   england  ...          NaN             0
45  K02000001            united kingdom  ...          NaN             0
46  K03000001             great britain  ...          NaN             0
47  K04000001         england and wales  ...          NaN             0
48  N92000002          northern ireland  ...          NaN             0
49  S92000003                  scotland  ...          NaN             0
50  W92000004                     wales  ...          NaN             0

[51 rows x 12 columns]

 Exercise 4 - prepare dataset for normalisation 
---
1. Read Covid vaccination data from the `by_country` sheet in the Excel file at this link : https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/`Covid`%20Vaccination%20Data.xlsx?raw=true
2. Use .info() to find out which columns have missing values
3. Remove all rows with missing data in the total_vaccination column
4. Find the median vaccinations per hundred 
5. display the mean people vaccinated per hundred for each country in descending order
6. find the range of total_vaccinations across the dataframe 


**Test output**:  
1. dataframe is saved in a variable
2. 
```
RangeIndex: 14994 entries, 0 to 14993
Data columns (total 15 columns):
    Column                               Non-Null Count  Dtype         
                                
 0   country                              14994 non-null  object        
 1   iso_code                             14994 non-null  object        
 2   date                                 14994 non-null  datetime64[ns]
 3   total_vaccinations                   9011 non-null   float64       
 4   people_vaccinated                    8370 non-null   float64       
 5   people_fully_vaccinated              6158 non-null   float64       
 6   daily_vaccinations_raw               7575 non-null   float64       
 7   daily_vaccinations                   14796 non-null  float64       
 8   total_vaccinations_per_hundred       9011 non-null   float64       
 9   people_vaccinated_per_hundred        8370 non-null   float64       
 10  people_fully_vaccinated_per_hundred  6158 non-null   float64       
 11  daily_vaccinations_per_million       14796 non-null  float64       
 12  vaccines                             14994 non-null  object        
 13  source_name                          14994 non-null  object        
 14  source_website                       14994 non-null  object        
dtypes: datetime64[ns](1), float64(9), object(5)
memory usage: 1.7+ MB
```
3. 9011 rows × 15 columns
4. 7.78
5. 
```
country
Gibraltar       72.172462
Maldives        51.276087
Israel          48.931008
Seychelles      48.233333
Aruba           42.155000
                  ...    
Guinea           0.677000
Sierra Leone     0.530000
Namibia          0.317143
South Africa     0.261667
Albania          0.080000
Name: people_vaccinated_per_hundred, Length: 104, dtype: float64
```
6. 275338000.0


Exercise 4.1


In [6]:
#importing libraries
import pandas as pd
#reading the data set from an excel file using the Pandas Library
df = pd.read_excel('https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true')
#Renaming the variable
dfvac = df
dfvac

country  ...                                     source_website
0      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
1      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
2      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
3      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
4      Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
...            ...  ...                                                ...
14989     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14990     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14991     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14992     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...
14993     Zimbabwe  ...  https://twitter.com/MoHCCZim/status/1388935941...

[14994 rows x 15 columns]

Exercise 4.2


In [7]:
#Use .info() to find out which columns have missing values
dfvac.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14994 entries, 0 to 14993
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   country                              14994 non-null  object        
 1   iso_code                             14994 non-null  object        
 2   date                                 14994 non-null  datetime64[ns]
 3   total_vaccinations                   9011 non-null   float64       
 4   people_vaccinated                    8370 non-null   float64       
 5   people_fully_vaccinated              6158 non-null   float64       
 6   daily_vaccinations_raw               7575 non-null   float64       
 7   daily_vaccinations                   14796 non-null  float64       
 8   total_vaccinations_per_hundred       9011 non-null   float64       
 9   people_vaccinated_per_hundred        8370 non-null   float64       
 10  people_ful

Exercise 4.3


In [13]:
df = pd.read_excel('https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true')
#Remove all rows with missing data in the total_vaccination column
dfvac.dropna(subset = ['total_vaccinations'])
display(dfvac.head())
display(dfvac.info())

country  ...                                     source_website
0  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
1  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
2  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
3  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...
4  Afghanistan  ...  https://reliefweb.int/report/afghanistan/afgha...

[5 rows x 15 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14994 entries, 0 to 14993
Data columns (total 15 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   country                              14994 non-null  object        
 1   iso_code                             14994 non-null  object        
 2   date                                 14994 non-null  datetime64[ns]
 3   total_vaccinations                   9011 non-null   float64       
 4   people_vaccinated                    8370 non-null   float64       
 5   people_fully_vaccinated              6158 non-null   float64       
 6   daily_vaccinations_raw               7575 non-null   float64       
 7   daily_vaccinations                   14796 non-null  float64       
 8   total_vaccinations_per_hundred       9011 non-null   float64       
 9   people_vaccinated_per_hundred        8370 non-null   float64       
 10  people_ful

None

Exercise 4.4


In [10]:
df = pd.read_excel('https://github.com/lilaceri/Working-with-data-/blob/342abab10d93c4bf23b5c55a50f189f12a137c5f/Data%20Sets%20for%20code%20divisio/Covid%20Vaccination%20Data.xlsx?raw=true')
# Python code to show the working of median() function.
dfvac2["total_vaccinations_per_hundred"].median()

NameError: ignored

Exercise 4.5


In [ ]:
dfvac3 = dfvac2.dropna(subset = ['people_vaccinated_per_hundred'])

dfvac3.groupby(['country'])['people_vaccinated_per_hundred',].mean().sort_values(['people_vaccinated_per_hundred'],ascending = False)

Exercise 5 - Correcting the format of dates where necessary